### PR TITLE
"Back My Stack" with anonymous profile

### LINKED ISSUE #182: Ability to "Back My Stack" with an anonymous profile
### Story

As anonymous user, I want to upload my dependency files and "Back My Stack" using the special Open Collective Tier.

### Scenario

1. Go on https://backyourstack.com/
2. Upload a first dependency file, ie: https://raw.githubusercontent.com/opencollective/opencollective-frontend/master/package.json
3. Upload a first dependency file, ie: https://raw.githubusercontent.com/opencollective/opencollective-api/master/package.json
4. Click on a "Save Profile" button
 - The files will be persisted to s3 under a randomly generated identifier, ie: `Pi9vF3knlTIwKcFtmBvzX4NHGmAfbeNr`
 - The `backing.json`will be available at https://backyourstack.com/Pi9vF3knlTIwKcFtmBvzX4NHGmAfbeNr/backing.json
5. Click on a "Back My Stack" button
6. A confirmation page will explain in plain english what "Backing My Stack" means, the main action is a button that
redirect to http://staging.opencollective.com/backyourstack2/contribute/level-1-9912/checkout?data={%22jsonUrl%22:%22http:\/\/backyourstack.com\/Pi9vF3knlTIwKcFtmBvzX4NHGmAfbeNr\/backing.json%22}
6. On success from the Open Collective flow, I'm redirected on a BackYourStack page confirming what happened and what I can do next

### PR DESCRIPTION
**Pull request by flickz**
This PR attempt to fix #182 Currently saves the package and displays the file url in a non editable input.
[Updated]
### Screen record
dispatcher

### CODE REVIEW

File: src/next.config.js [hunk 1]
```diff
@@ -58,4 +58,7 @@ module.exports = withCSS({
 
     return config;
   },
+  publicRuntimeConfig: {
+    openCollectiveRedirectUrl: process.env.OPENCOLLECTIVE_REDIRECT_URL,
```
Comment (znarf) [inline src/next.config.js]: This might be a valid solution, but we should not have 2 different ways of doing the same thing. 

So please process `OPENCOLLECTIVE_REDIRECT_URL` like `OPENCOLLECTIVE_REFERRAL`

File: src/next.config.js [hunk 2]
```diff
@@ -17,6 +17,7 @@ module.exports = withCSS({
       // Make some environment variables accessible from the client
       new webpack.EnvironmentPlugin({
         OPENCOLLECTIVE_REFERRAL: null,
+        OPENCOLLECTIVE_REDIRECT_URL: process.env.OPENCOLLECTIVE_REDIRECT_URL,
```
Comment (znarf) [inline src/next.config.js]: Was that tested?

Comment (znarf) [reply src/next.config.js]: Sorry, I had `OPENCOLLECTIVE_REDIRECTION_URL` instead of `OPENCOLLECTIVE_REDIRECT_URL`.

Anyway, here, just put `null` like `OPENCOLLECTIVE_REFERRAL`, this is the default value.

File: now.json [hunk 3]
```diff
@@ -9,7 +9,11 @@
     "GITHUB_TOKEN_DONATORS": "@github_token_donators",
     "MAILCHIMP_API_KEY": "@mailchimp_api_key",
     "MAILCHIMP_API_URL": "@mailchimp_api_url",
-    "OPENCOLLECTIVE_REFERRAL": "20206"
+    "OPENCOLLECTIVE_REFERRAL": "20206",
+    "AWS_ACCESS_KEY_ID": "@aws_access_key_id",
+    "AWS_SECRET_KEY": "@aws_secret_key",
+    "AWS_S3_BUCKET": "@aws_s3_bucket",
+    "OPENCOLLECTIVE_REDIRECT_URL": "@opencollective_redirection_url"
```
Comment (znarf) [inline now.json]: The names needs to match, otherwise it's confusing.

File: src/pages/files.js [hunk 4]
```diff
@@ -88,112 +130,204 @@ export default class Files extends React.Component {
...
+  };
+
+  handleContinueBacMyStack = () => {
+    // Get the key url of the file
+    const { savedFileUrl } = this.state;
+    const { openCollectiveRedirectUrl } = this.props;
+    const { hostname, protocol } = window.location;
+    const fileKey = savedFileUrl.Key.replace('.json', '');
+    const jsonUrl = `"${protocol}//${hostname}/${fileKey}/file/backing.json"`;
+    const contributionUrl = `${openCollectiveRedirectUrl}?data={"jsonUrl":${jsonUrl}}`;
```
Comment (znarf) [inline src/pages/files.js]: This is not properly constructed.

- a JSON is constructed with JSON.stringify()
- URL query parameters needs to be properly handled, for example with the URL API: https://developer.mozilla.org/en-US/docs/Web/API/URL

**Comment by znarf** [comment]
Single file upload should be handled exactly like multiple file upload. Having 2 strategies is introducing complexity for no reason. A single file is just an array of length one :-)

Also, let's review the way files are stored in the S3 bucket.

Let's say that we generated such identifier: `0ba83370-a329-11e9-a4bb-65d900563d2d`.

The meta-data file should then be stored at: `0ba83370-a329-11e9-a4bb-65d900563d2d/dependencies.json`.

And it should be in this form:

```
{files: {"backyourstack":{"url":"https://opencollective-backyourstack-test.s3.amazonaws.com/0ba83370-a329-11e9-a4bb-65d900563d2d/backyourstack.json"},"opencollective-frontend":{"url":"https://opencollective-backyourstack-test.s3.amazonaws.com/0ba83370-a329-11e9-a4bb-65d900563d2d/opencollective-frontend.json"}}}
```

Adding a top level `files` will ensure we can store more things in the future in this structure without changing it.

The files should be prefixed with the identifier, so we emulate a folder.

File: src/pages/files.js [hunk 5]
```diff
@@ -88,112 +130,207 @@ export default class Files extends React.Component {
...
+    // Get the key url of the file
+    const { savedFileUrl } = this.state;
+    const { openCollectiveRedirectUrl } = this.props;
+    const { hostname, protocol } = window.location;
+    const uuid = savedFileUrl.Key.split('/')[0];
+    const jsonUrl = `${protocol}//${hostname}/${uuid}/file/backing.json`;
+    const data = JSON.stringify({
+      jsonUrl,
+    });
+    const contributionUrl = `${openCollectiveRedirectUrl}?data={${data}}`;
```
Comment (znarf) [inline src/pages/files.js]: This stills looks not correct.

- Looks like the JSON will have duplicate braces `{${data}}`
- URL query parameters (here `data`) needs to be properly handled, for example with the URL API: https://developer.mozilla.org/en-US/docs/Web/API/URL

Comment (flickz) [reply src/pages/files.js]: Noticed already, on it

File: src/next.config.js (hunk 2 above)
Comment (znarf) [reply src/next.config.js]: Don't forget that.

File: src/pages/files.js [hunk 6]
```diff
@@ -136,18 +143,20 @@ export default class Files extends React.Component {
...
-    const contributionUrl = `${openCollectiveRedirectUrl}?data={${data}}`;
-    window.location.replace(contributionUrl);
+    const { openCollectiveRedirectUrl, baseUrl } = this.props;
+    if (savedFileUrl) {
+      const uuid = savedFileUrl.Key.split('/')[0];
+      const jsonUrl = `${baseUrl}/${uuid}/file/backing.json`;
+      const data = JSON.stringify({
+        jsonUrl,
+      });
+      const searchParams = new URLSearchParams(`data=${data}`);
```
Comment (znarf) [inline src/pages/files.js]: Why not new `URLSearchParams({data });`?

Comment (flickz) [reply src/pages/files.js]: The generated string with that appears in compete, generate something like this
```
%7B%22jsonUrl%22%3A%22http%3A%2F%2Flocalhost%3A3001%2Fb0baf030-a3da-11e9-a530-1ddec81b7392%2Ffile%2Fbacking.json%22%7D%3D

{"jsonUrl":"http://localhost:3001/b0baf030-a3da-11e9-a530-1ddec81b7392/file/backing.json"}=
```

Comment (znarf) [reply src/pages/files.js]: Reading the documentation, it might be: `new URLSearchParams(['data', data])`

https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams/URLSearchParams

Comment (znarf) [reply src/pages/files.js]: `new URLSearchParams({ data })`works great for me

Comment (flickz) [reply src/pages/files.js]: I was doing something wrong, you're right it works.

File: src/server/index.js [hunk 7]
```diff
@@ -216,6 +233,47 @@ nextApp.prepare().then(() => {
...
+          opencollectiveAccount &&
+          get(opencollectiveAccount, 'orders.nodes', []).find(
+            order =>
+              opencollective && opencollective.slug === order.toAccount.slug,
+          );
+        if (order) {
+          opencollective.order = order;
+        }
+        return {
+          weigh: 100,
```
Comment (znarf) [inline src/server/index.js]: Let's fix the `weigh` typo everywhere please.

File: src/pages/confirmed.js [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    this.state = {
+      dispatchedOrders: [],
+    };
+  }
+
+  async componentDidMount() {
+    if (!this.props.orderId) {
+      return;
+    }
+    const order = await fetchOrder(parseInt(this.props.orderId));
```
Comment (znarf) [inline src/pages/confirmed.js]: Instead of fetching the order here, we should execute the `dispatchOrder` mutation from there and collect the result.

**Comment by znarf** [comment]
Why do we have now `OPENCOLLECTIVE_URL` instead of `OPENCOLLECTIVE_REDIRECT_URL` ?

This is breaking the deployment because the variable is not set anymore.

**Comment by flickz** [comment]
I updated environment variables because i want the `baseUrl` of `/api/graphql/` to be configurable in order to test the `dispatchOrder` mutation locally. 
Changes are
```
OPENCOLLECTIVE_REDIRECT_PATH=/brusselstogether/contribute/backyourstack-2817/checkout
OPENCOLLECTIVE_BASE_URL=http://localhost:3000
```

File: src/next.config.js [hunk 9]
```diff
@@ -17,6 +17,8 @@ module.exports = withCSS({
       // Make some environment variables accessible from the client
       new webpack.EnvironmentPlugin({
         OPENCOLLECTIVE_REFERRAL: null,
+        OPENCOLLECTIVE_REDIRECT_PATH: process.env.OPENCOLLECTIVE_REDIRECT_PATH,
+        OPENCOLLECTIVE_BASE_URL: process.env.OPENCOLLECTIVE_BASE_URL,
```
Comment (znarf) [inline src/next.config.js]: We don't need default values, it's the job of the plugin to extract them.

See: https://webpack.js.org/plugins/environment-plugin/

File: src/next.config.js [hunk 10]
```diff
@@ -17,6 +17,8 @@ module.exports = withCSS({
       // Make some environment variables accessible from the client
       new webpack.EnvironmentPlugin({
         OPENCOLLECTIVE_REFERRAL: null,
+        OPENCOLLECTIVE_REDIRECT_PATH: process.env.OPENCOLLECTIVE_REDIRECT_PATH,
```
Comment (znarf) [inline src/next.config.js]:
```suggestion
        OPENCOLLECTIVE_REDIRECT_PATH: null,
```

File: src/next.config.js (hunk 9 above)
Comment (znarf) [inline src/next.config.js]:
```suggestion
        OPENCOLLECTIVE_BASE_URL: null,
```

File: src/pages/backmystack.js [hunk 11]
```diff
@@ -0,0 +1,119 @@
+import React, { Fragment } from 'react';
+import PropTypes from 'prop-types';
+
+import Header from '../components/Header';
+import Footer from '../components/Footer';
+
+export default class BackMyStack extends React.Component {
+  static getInitialProps({ req, query }) {
+    const uuid = query.uuid; // handle case where the uuid is not supplied
+    const opencollectiveBaseUrl = process.env.OPENCOLLECTIVE_BASE_URL;
```
Comment (znarf) [inline src/pages/backmystack.js]: The convention accross our projects it to name this variable `WEBSITE_URL`. Maybe, we can name it `OPENCOLLECTIVE_WEBSITE_URL` here.

File: src/lib/opencollective.js [hunk 12]
```diff
@@ -2,8 +2,9 @@ import request from 'graphql-request';
 
 import cache from './cache';
 
-const baseUrl = 'https://opencollective.com/api/graphql';
-const baseUrlV2 = 'https://opencollective.com/api/graphql/v2';
+const opencollectiveBaseUrl = process.env.OPENCOLLECTIVE_BASE_URL;
```
Comment (znarf) [inline src/lib/opencollective.js]: The convention accross our projects is to name this variable `API_URL`, maybe we can name it `OPENCOLLECTIVE_API_URL` here. The `/api/` path should then be removed.

File: src/components/Modal.js [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+    children: PropTypes.node,
+  };
+
+  constructor(props) {
+    super(props);
+  }
+
+  render() {
+    if (typeof document !== 'undefined')
+      document.body.classList[this.props.show ? 'add' : 'remove']('showModal');
```
Comment (znarf) [inline src/components/Modal.js]: I'm surprised we don't have a lint rule for that (or prettier), anyway, put braces here.

Background: https://www.joelonsoftware.com/2005/05/11/making-wrong-code-look-wrong/

File: src/pages/backmystack.js [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+    const openCollectiveRedirectUrl = `${opencollectiveBaseUrl}${process.env.OPENCOLLECTIVE_REDIRECT_PATH}`;
+
+    let protocol = 'https:';
+    const host = req ? req.headers.host : window.location.hostname;
```
Comment (znarf) [inline src/pages/backmystack.js]: When I tested, the port was missing. Not sure if it was the case with or without `req`.

`window.location.host` perhaps?

**Comment by znarf** [comment]
@flickz have you noticed https://github.com/opencollective/backyourstack/pull/194

Closing this PR in favor of it

**Comment by flickz** [comment]
Oh great
